src/coreclr/jit/fgbasic.cpp

```diff
@@ -516,6 +562,155 @@ void Compiler::fgReplaceSwitchJumpTarget(BasicBlock* blockSwitch, BasicBlock* ne
     noway_assert(!"Did not find oldTarget in jumpTab[]");
 }
 
+//------------------------------------------------------------------------
+// fgChangeEhfBlock: We have a BBJ_EHFINALLYRET block at 'oldBlock' and we want to move this
+// to 'newBlock'. All of the 'oldBlock' successors need to have their predecessor lists updated
+// by removing edges to 'oldBlock' and adding edges to 'newBlock'.
+//
+// Arguments:
+//   oldBlock - previous BBJ_EHFINALLYRET block
+//   newBlock - block that is replacing 'oldBlock'
+//
+void Compiler::fgChangeEhfBlock(BasicBlock* oldBlock, BasicBlock* newBlock)
+{
+    assert(oldBlock != nullptr);
+    assert(newBlock != nullptr);
+    assert(oldBlock->KindIs(BBJ_EHFINALLYRET));
+    assert(fgPredsComputed);
+
+    for (BasicBlock* const succ : oldBlock->EHFinallyRetSuccs())
+    {
+        assert(succ != nullptr);
+
+        // Remove the old edge [oldBlock => succ]
+        //
+        assert(succ->countOfInEdges() > 0);
+        fgRemoveRefPred(succ, oldBlock);
+
+        // Create the new edge [newBlock => succ]
+        //
+        fgAddRefPred(succ, newBlock);
+    }
+}
+
+//------------------------------------------------------------------------
+// fgReplaceEhfSuccessor: update BBJ_EHFINALLYRET block so that all control
+//   that previously flowed to oldSucc now flows to newSucc. It is assumed
+//   that oldSucc is currently a successor of `block`.
+//
+// Arguments:
+//   block   - BBJ_EHFINALLYRET block
+//   oldSucc - old successor
+//   newSucc - new successor
+//
+void Compiler::fgReplaceEhfSuccessor(BasicBlock* block, BasicBlock* oldSucc, BasicBlock* newSucc)
+{
+    assert(block != nullptr);
+    assert(oldSucc != nullptr);
+    assert(newSucc != nullptr);
+    assert(block->KindIs(BBJ_EHFINALLYRET));
+    assert(fgPredsComputed);
+
+    BBehfDesc* const ehfDesc   = block->GetJumpEhf();
+    unsigned         succCount = ehfDesc->bbeCount;
+    BasicBlock**     succTab   = ehfDesc->bbeSuccs;
+
+    unsigned i = 0;
+
+    // Walk the successor table looking for blocks to update the preds for
+    while (i < succCount)
+    {
+        if (succTab[i] == oldSucc) // We will update when succTab[i] matches
+        {
+            // Remove the old edge [block => oldSucc]
+            //
+            fgRemoveAllRefPreds(oldSucc, block);
+
+            // Change the succTab entry to branch to the new location
+            //
+            succTab[i] = newSucc;
+
+            // Create the new edge [block => newSucc]
+            //
+            FlowEdge* const newEdge = fgAddRefPred(newSucc, block);
+
+            // Now set the correct value of newEdge's DupCount
```

Comment:
Seems like this could be a bit simpler if you remember that you'd already created a flow edge, something like

```c++
FlowEdge* edge = nullptr;
for (i = 0; i < succCount; i++)
{
    if (succTab[i] != oldSucc) continue;

    succTab[i] = newSucc;

    if (flowEdge == nullptr)
    {
       flowEdge = fgAddRefPred(newSucc, block);
    }
    else
    {
       flowEdge->incrementDupCount();
    }
}
```